ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/validation/RequestValidator.java

```diff
@@ -0,0 +1,145 @@
+    this.snapshotSupplier = snapshotSupplier;
+    this.registryClient = registryClient;
+    this.ksqlConfig = ksqlConfig;
+  }
+
+  /**
+   * Validates the messages against a snapshot in time of the KSQL engine.
+   *
+   * @param statements          the list of statements to validate
+   * @param propertyOverrides   a map of properties to override for this validation
+   * @param sql                 the sql that generated the list of statements, used for
+   *                            generating more useful debugging information
+   *
+   * @throws KsqlException if any of the statements cannot be validated, or the number
+   *                       of requested statements would cause the execution context
+   *                       to exceed the number of persistent queries that it was configured
+   *                       to support
+   */
+  public void validate(
+      final List<ParsedStatement> statements,
+      final Map<String, Object> propertyOverrides,
+      final String sql
+  ) {
+    final EngineSnapshot snapshot = snapshotSupplier.get();
+
+    int numPersistentQueries = 0;
+    for (ParsedStatement statement : statements) {
+      numPersistentQueries += validate(statement, ksqlConfig, propertyOverrides, snapshot);
+    }
+
+    final KsqlExecutionContext ctx = snapshot.context;
+    if (QueryCapacityUtil.exceedsPersistentQueryCapacity(ctx, ksqlConfig, numPersistentQueries)) {
+      QueryCapacityUtil.throwTooManyActivePersistentQueriesException(ctx, ksqlConfig, sql);
+    }
+  }
+
+  /**
+   * @return the number of persistent queries that were validated
+   *
+   * @throws KsqlException if the statement cannot be validated
+   */
+  @SuppressWarnings("unchecked")
+  private int validate(
+      final ParsedStatement parsed,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> propertyOverrides,
+      final EngineSnapshot snapshot
+  ) throws KsqlException  {
+    final PreparedStatement prepared = snapshot.context.prepare(parsed);
+    final Statement statement = prepared.getStatement();
+    final Class<? extends Statement> statementClass = statement.getClass();
+    final StatementValidator customValidator = customValidators.get(statementClass);
+
+    if (customValidator != null) {
+      customValidator.validate(prepared, snapshot.context, ksqlConfig, propertyOverrides);
+    } else if (KsqlEngine.isExecutableStatement(prepared)) {
+      snapshot.context.execute(
+          StatementWithSchema.forStatement(prepared, registryClient),
+          ksqlConfig,
+          propertyOverrides
+      );
+    } else if (!optionalToValidate.contains(prepared.getStatement().getClass())) {
+      throw new KsqlStatementException(
+          "Do not know how to validate statement of type: " + statementClass,
+          prepared.getStatementText());
+    }
+
+    return (statement instanceof CreateAsSelect || statement instanceof InsertInto) ? 1 : 0;
+  }
+
+  public static class EngineSnapshot {
```

Comment:
Can we drop this for now? You can add back in a later PR that actually adds functionality to this.